common-kafka/src/main/java/com/cerner/common/kafka/consumer/ProcessingKafkaConsumer.java

```diff
@@ -755,17 +755,19 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitionsAssigned)
             LOGGER.info("Rebalance [{}] is in progress. Old partitions: {} New Partitions: {}", REBALANCE_COUNTER.count(),
                     partitions.keySet(), partitionsAssigned);
 
-            // Close all partitions we will no longer use
+            /**
+             * We must reset the local partition state (cached offsets) for all partitions,
+             * even if the last assignment we saw also included these partitions,
+             * because we may have been disconnected and missed an entire group generation
+             * during which another consumer might have been assigned the partition and committed offsets.
+             */
             partitions.values().stream()
-                    .filter(partition -> !partitionsAssigned.contains(partition.getTopicPartition()))
                     .forEach(partition -> IOUtils.closeQuietly(partition));
 
-            // Remove all partitions we are no longer assigned to
-            partitions.keySet().retainAll(partitionsAssigned);
+            partitions.clear();
 
-            // Add all partitions that we don't already have
             try {
-                partitionsAssigned.stream().filter(tp -> !partitions.containsKey(tp))
+                partitionsAssigned.stream()
                         .forEach(tp -> partitions.put(tp, buildPartition(tp, config, consumer)));
```

Comment:
Since the `consumer` is left alone, are we sure that its internal state (i.e. fetch cursors) is consistent with what's being done here? If so I'm not sure there is any real benefit to proactively recreating all of the `ProcessingPartition`s. They'll be initialized eventually when we first retrieve a message for them, which is the same behavior as when a new `ProcessingKafkaConsumer` is created. And if not, we might need to call `partition.resetToCommitted()` for each partition. I'm not sure how Kafka handles state management for partitions that stay assigned to the same consumer group member.

Reply:
> Since the consumer is left alone, are we sure that its internal state (i.e. fetch cursors) is consistent with what's being done here?

The fetch cursors don't need to be in sync with the ProcessingPartition commit offsets, right?

> If so I'm not sure there is any real benefit to proactively recreating all of the ProcessingPartitions. They'll be initialized eventually when we first retrieve a message for them, which is the same behavior as when a new ProcessingKafkaConsumer is created.

The ProcessingPartitions.lastCommittedOffset offset field is initialized at ProcessingPartition construction time, not when we first retrieve a message.

If a ProcessingPartition was constructed in a previous generation (before the rebalances) then it's lastCommittedOffset would not reflect commits made by other consumers while we were disconnected from the group.

> And if not, we might need to call partition.resetToCommitted() for each partition.

resetToCommitted() won't help us here, because lastCommittedOffset is corrupted / out of sync with the current group generation.
And since it also attempts to seek the consumer to this invalid lastCommittedOffset, I'm not sure what the result would be.

https://github.com/cerner/common-kafka/blob/master/common-kafka/src/main/java/com/cerner/common/kafka/consumer/ProcessingPartition.java#L491-L497

To avoid the additional object construction, we could also add a method to ProcessingPartition to refresh with the last committed offset from remote. Basically a public method to do 
lastCommittedOffset = getLastCommittedOffset()

That would avoid the partition getting stuck unable to commit, but I think the rest of the state of the ProcessingPartition would also be invalid and need to be cleared.

Assuming the partition was processing messages before the first rebalance happened, then it might have recentProcessingResults, pendingOffsets, and completedOffsets which would not make sense with the current generation, right?

> I'm not sure how Kafka handles state management for partitions that stay assigned to the same consumer group member.

I'm not 100% sure, but I think it validates the cursor state by fetching offsets from the server, in the state transition methods that are invoked when the rebalance listener is triggered.

If I'm understanding correctly, if the Fetcher failed to update the cursors to reflect offsets committed by a different consumer, then the consequence would be duplicate message processing, not deadlock, right?

Reply:
I think there is a scenario where retaining some of the ProcessingPartition state across a rebalance might have a benefit.

Technically, the whole state isn't invalid, it just needs to be adjusted to reflect the activity other consumers may have committed to the partitions during any generations we missed.

If before the first rebalance we successfully processed messages and acked them, but didn't get to commit them, then we don't really need to reprocess those messages. We would just want to clear the offset lists of any offsets earlier than the true last committed offset.

I'm not sure if this scenario is likely to actually happen though, and I'm debating whether it's worth adding the complexity to the implement a new "reset" method to ProcessingPartition which tries to stitch together a valid state after a rebalance.

The only benefit of this would be to reduce duplicate message processing after a rebalance, and the means of doing this feels very brittle to me.

Reply:
> I'm not sure if this scenario is likely to actually happen though

Thinking more about this, I'm pretty sure this would be the most common case on rebalance. There would be some completed, pending, and committable offsets in the ProcessingPartition, and only in the missed generation case would these need to be pruned.

I'm debating whether it would be better to make the more invasive code change (pruning offset state in ProcessingPartition with a new public method on rebalance), to avoid making a more invasive change in the client's behavior (increasing duplicate processing after a rebalance, in the common case).

Reply:
I'm fine with either solution as long as we have high confidence from the unit/integration tests. Rebalancing should be fairly rare so we don't necessarily need to optimize for that, but if it is simpler overall or safer in terms of the amount of behavior that is changing then that alternative approach may be preferable.

Reply:
That sounds reasonable to me. 

If there isn't a strong reason to pick either approach, I'm going to go with KISS and just reset the ProcessingPartition state, as we have in this PR.